worker: Support restarting a failed streaming query.

michael-berlin opened the pull request:
@alainjobart 

We'll restart the query by advancing the start of the chunk right after the primary key of the last row read so far.

Other changes:
- Refactored FindChunks. I'm using sqltypes.Result now.
- Added Chunk struct to store a chunk. I'm using sqltypes.Value for the start and end to simplify processing the values.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/youtube/vitess/1845)
<!-- Reviewable:end -->

alainjobart commented:
Maybe add a test for restart with multiple keys...

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[go/vt/worker/restartable_result_reader.go, line 157 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1845#-KLqgWMBMW_2FbjTv3jl:-KLqgWMCiX70ANRh06NW:398759654) ([raw file](https://github.com/youtube/vitess/blob/ed7f16cb1020985b35c0328cf6822025fa7561a6/go/vt/worker/restartable_result_reader.go#L157)):*
> ```Go
> 		// string types and MySQL collation rules.
> 		var b bytes.Buffer
> 		for i := 0; i < len(r.td.PrimaryKeyColumns); i++ {
> ```

I think this is wrong, only works for one column? For multiple columns, let's say a and b, it seems you will produce: 'a>10b>10', it needs to be: 'a>10 or (a=10 and b>10)'. Fun stuff.

I would do it as a separate method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1845#-:-KLqhLBbhEdJNazWv6Xl:-876490391)*
<!-- Sent from Reviewable.io -->

alainjobart commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1845#-:-KLuGk136WN1hrWqjSIw:1417088721)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[go/vt/worker/restartable_result_reader.go, line 157 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1845#-KLqgWMBMW_2FbjTv3jl:-KLuH_OQOIQ7VEBemQYS:1220955376) ([raw file](https://github.com/youtube/vitess/blob/ed7f16cb1020985b35c0328cf6822025fa7561a6/go/vt/worker/restartable_result_reader.go#L157)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> I think this is wrong, only works for one column? For multiple columns, let's say a and b, it seems you will produce: 'a>10b>10', it needs to be: 'a>10 or (a=10 and b>10)'. Fun stuff.
> 
> I would do it as a separate method.
</details>
Done.

I've moved the code into a separate method and added unit tests for it.

I'm also using the following short-form now which simplified the code: (a,b,c) > (1,2,3)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1845)*
<!-- Sent from Reviewable.io -->
